internal/overlord/logstate/manager_test.go

```diff
@@ -0,0 +1,246 @@
+			"svc2": {LogTargets: []string{}},
+			"svc3": {LogTargets: []string{"unset"}},
+			"svc4": {LogTargets: []string{"optout"}},
+			"svc5": {LogTargets: []string{"optin"}},
+			"svc6": {LogTargets: []string{"disabled"}},
+			"svc7": {LogTargets: []string{"unset", "optin", "disabled"}},
+		},
+	}
+
+	expected := map[string]map[string]*plan.LogTarget{
+		"svc1": {"unset": &unset, "optout": &optout},
+		"svc2": {"unset": &unset, "optout": &optout},
+		"svc3": {"unset": &unset},
+		"svc4": {"optout": &optout},
+		"svc5": {"optin": &optin},
+		"svc6": {},
+		"svc7": {"unset": &unset, "optin": &optin},
+	}
+
+	logbuf, restore := logger.MockLogger("")
+	defer restore()
+
+	planTargets := selectTargets(&input)
+	c.Check(planTargets, DeepEquals, expected)
+	// Check no error messages were logged
+	c.Check(logbuf.Bytes(), HasLen, 0)
+}
+
+func (s *managerSuite) TestLogManager(c *C) {
+	m := NewLogManager()
+	// Fake ringbuffer so that log manager can create forwarders
+	rb := servicelog.RingBuffer{}
+
+	// Start 3 services
+	var wg sync.WaitGroup
+	wg.Add(4)
+
+	// Call PlanChanged with new plan
+	go func() {
+		defer wg.Done()
+		m.PlanChanged(&plan.Plan{
+			Services: map[string]*plan.Service{
+				"svc1": {},
+				"svc2": {LogTargets: []string{"optin", "disabled"}},
+				"svc3": {LogTargets: []string{"unset"}},
+			},
+			LogTargets: map[string]*plan.LogTarget{
+				"unset":    {Type: plan.LokiType, Selection: plan.UnsetSelection},
+				"optout":   {Type: plan.LokiType, Selection: plan.OptOutSelection},
+				"optin":    {Type: plan.LokiType, Selection: plan.OptInSelection},
+				"disabled": {Type: plan.LokiType, Selection: plan.DisabledSelection},
+			},
+		})
+	}()
+	// Ensure PlanChanged gets the lock first
+	time.Sleep(1 * time.Millisecond)
+
+	// Call ServiceStarted
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc1", &rb)
+	}()
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc2", &rb)
+	}()
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc3", &rb)
+	}()
+
+	wg.Wait()
+	c.Check(m.forwarders, HasLen, 4)
+	checkForwarderExists(c, m.forwarders, "svc1", "unset")
+	checkForwarderExists(c, m.forwarders, "svc1", "optout")
+	checkForwarderExists(c, m.forwarders, "svc2", "optin")
+	checkForwarderExists(c, m.forwarders, "svc3", "unset")
+
+	// Update the plan
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		m.PlanChanged(&plan.Plan{
+			Services: map[string]*plan.Service{
+				"svc1": {},
+				"svc2": {LogTargets: []string{"optout", "disabled"}},
+				"svc4": {LogTargets: []string{"optin"}},
+			},
+			LogTargets: map[string]*plan.LogTarget{
+				"unset":    {Type: plan.LokiType, Selection: plan.UnsetSelection},
+				"optout":   {Type: plan.LokiType, Selection: plan.OptOutSelection},
+				"optin":    {Type: plan.LokiType, Selection: plan.OptInSelection},
+				"disabled": {Type: plan.LokiType, Selection: plan.DisabledSelection},
+			},
+		})
+	}()
+	// Ensure PlanChanged gets the lock first
+	time.Sleep(1 * time.Millisecond)
```

Comment:
Again, we need to synchronize this rather than `time.Sleep` to avoid racy tests.